globalsat_sport.cc

```diff
@@ -0,0 +1,901 @@
+	}
+
+	if (payload != NULL) {
+		for (i = 0; i < size; i++) {
+			write_byte(payload[i]);
+			crc ^= payload[i];
+		}
+	}
+	write_byte(crc);
+	if (global_opts.debug_level > 1) {
+		printf("\n");
+	}
+}
+
+static uint8_t *
+globalsat_read_package(int *out_length, uint8_t * out_DeviceCommand)
+{
+	uint8_t DeviceCommand, len_h, len_l, crc;
+	uint8_t *payload;
+	int length;
+	int i;
+	uint8_t calc_crc = 0;
+
+	DeviceCommand = recv_byte();
+	if (global_opts.debug_level > 1) {
+		printf("DeviceCommand: 0x%02x ", DeviceCommand);
+	}
+	len_h = recv_byte();
+	calc_crc ^= len_h;
+	len_l = recv_byte();
+	calc_crc ^= len_l;
+
+	length = (len_h << 8) + len_l;
+	if (global_opts.debug_level > 1) {
+		printf("len=%d Payload:", length);
+	}
+
+	payload = (uint8_t *) malloc(length);
+	if (payload == NULL) {
+		goto error_out;
+	}
+
+	for (i = 0; i < length; i++) {
+		payload[i] = recv_byte();
+		calc_crc ^= payload[i];
+	}
+
+	crc = recv_byte();
+	if (global_opts.debug_level > 1) {
+		printf("crc=0x%x should be=0x%x\n", crc, calc_crc);
+	}
+	if (crc == calc_crc) {
+		*out_DeviceCommand = DeviceCommand;
+		*out_length = length;
+		return payload;
+	}
+	//crc error
+	free(payload);
+	error_out:
+	*out_length = 0;
+	return NULL;
+}
+
+/*
+ * Send one byte package
+ */
+static void
+globalsat_send_simple(uint8_t command)
+{
+	uint8_t payload[1];
+	payload[0] = command;
+	globalsat_write_package(payload, 1);
+}
+
+static void
+globalsat_probe_device()
+{
+	//TODO try this first if fails try with false, to support 561
+	isSizeSwaped = FALSE;		//all devices but gh561 since gh561 has swaped size.
+
+	globalsat_send_simple(CommandWhoAmI);
+
+	uint8_t *payload;
+	int len;
+	uint8_t DeviceCommand;
+	payload = globalsat_read_package(&len, &DeviceCommand);
+	if ((len > 0) && (payload != NULL)) {
+		if (global_opts.debug_level > 1) {
+			printf("Got package!!!\n");
+		}
+		//TODO figure out what device it is if we start to support more devices then gh625XT
+	}
+
+	if (payload) {
+		free(payload);
+	}
+}
+
+static void
+rd_init(const char *fname)
+{
+	if (global_opts.debug_level > 1) {
+		printf(MYNAME " rd_init()\n");
+	}
+	if (opt_dump_file) {
+    	dumpfile = gbfopen(opt_dump_file, "w", MYNAME);
+    	if (!dumpfile) {
+			printf(MYNAME " rd_init() creating dumpfile %s FAILED continue anyway\n",opt_dump_file);
+    	} else {
+	    	if (global_opts.debug_level > 1) {
+   				printf(MYNAME " rd_init() creating dumpfile %s for writing binnary copy of serial stream\n",opt_dump_file);
+   			}
+		}
+    }
+    if (!opt_input_dump_file)
+    {
+		serial_init(fname);
+    } else {
+    	// read from dump-file instead of serial
+        in_file = gbfopen(fname, "r", MYNAME);
+        is_fatal(!in_file, "Could not open dumpfile for input: %s", fname);
+
+    }
+	globalsat_probe_device();
+}
+
+static void
+wr_init(const char *fname)
+{
+	if (global_opts.debug_level > 1) {
+		printf(MYNAME " wr_init()\n");
+	}
+	serial_init(fname);
+}
+
+
+static void
+rd_deinit(void)
+{
```

Comment:
There are some style nits, but overall, this is progressing nicely. Thank
you.

If you can find and address the leaks found by
https://travis-ci.org/gpsbabel/gpsbabel/jobs/102729158 to get a clean
presubmit, I'd welcome this into the next release. (Which is likely to be
this weekend.)

I've not checked, but I THINK it's the
  static char buf[80]
  sprintf(junk into buf)
  wpt->something = xstrdup(buf);

Either buf should be correctly sized by xasprintf and then freed or we
should just let QString do char* copy/duplication from the char* into the
QString and not introduce an additional malloc/copy via xstrdup.

I could be wrong, but that caught my eye during review, so it seems like a
bit of a smoking gun.

Thanx!
RJL

On Fri, Jan 15, 2016 at 11:36 PM, Robert Lipe <notifications@github.com>
wrote:

> In globalsat_sport.cc
> <https://github.com/gpsbabel/gpsbabel/pull/20#discussion_r49927646>:
>
> > +	globalsat_probe_device();
> > +}
> > +
> > +static void
> > +wr_init(const char *fname)
> > +{
> > +	if (global_opts.debug_level > 1) {
> > +		printf(MYNAME " wr_init()\n");
> > +	}
> > +	serial_init(fname);
> > +}
> > +
> > +
> > +static void
> > +rd_deinit(void)
> > +{
>
> I'll fix indention later.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gpsbabel/gpsbabel/pull/20/files#r49927646>.
>